File: languages/cpp/03-advanced/inheritance-and-polymorphism/example/main.cpp
// Module focus: Treating different concrete types through one common interface.
// Why it matters: practicing inheritance and polymorphism patterns makes exercises and checkpoints
// easier to reason about.

#include <iostream>
#include <memory>
#include <vector>
using namespace std;

class Shape {
  public:
    virtual ~Shape() = default;
    virtual double area() const = 0;
};

class Rectangle : public Shape {
  public:
    Rectangle(double widthValue, double heightValue) : width(widthValue), height(heightValue) {}

    double area() const override { return width * height; }

  private:
    double width;
    double height;
};

class Circle : public Shape {
  public:
    explicit Circle(double radiusValue) : radius(radiusValue) {}

    double area() const override { return 3.14159265358979323846 * radius * radius; }

  private:
    double radius;
};

// Walk through one fixed scenario so inheritance and polymorphism behavior stays repeatable.
int main() {
    // Prepare sample inputs that exercise the key inheritance and polymorphism path.
    vector<unique_ptr<Shape>> shapes;
    shapes.push_back(unique_ptr<Shape>(new Rectangle(3.0, 4.0)));
    shapes.push_back(unique_ptr<Shape>(new Circle(2.0)));

    for (const auto& shape : shapes) {
        // Report values so learners can verify the inheritance and polymorphism outcome.
        cout << "Area: " << shape->area() << '\n';
    }

    return 0;
}

File: languages/java/03-advanced/inheritance-and-polymorphism/example/Main.java
// Module focus: Treating different concrete types through one common interface.
// Why it matters: practicing inheritance and polymorphism patterns makes exercises and checkpoints easier to reason about.

import java.util.List;
import java.util.Locale;

public class Main {
    interface Shape {
        // The interface names the behavior generic code can rely on.
        double area();

        String name();
    }

    static class Rectangle implements Shape {
        private final double width;
        private final double height;

        Rectangle(double width, double height) {
            this.width = width;
            this.height = height;
        }

        @Override
        public double area() {
            // Each concrete class owns its formula behind the shared interface.
            return width * height;
        }

        @Override
        public String name() {
            return "Rectangle";
        }
    }

    static class Circle implements Shape {
        private final double radius;

        Circle(double radius) {
            this.radius = radius;
        }

        @Override
        public double area() {
            // Callers do not need to know this formula when they use Shape references.
            return Math.PI * radius * radius;
        }

        @Override
        public String name() {
            return "Circle";
        }
    }

    public static void main(String[] args) {
        Locale.setDefault(Locale.US);

        // Prepare sample inputs that exercise the key inheritance and polymorphism path.
        List<Shape> shapes = List.of(new Rectangle(3.0, 4.0), new Circle(2.0));

        for (Shape shape : shapes) {
            // Report values so learners can verify the inheritance and polymorphism outcome.
            System.out.printf("%s area: %.2f%n", shape.name(), shape.area());
        }
    }
}
